test/systemtest/utils.py

```diff
@@ -0,0 +1,428 @@
+            logging.debug("[File is too large to be shown ({} bytes).]".format(
+                f.stat().st_size))
+        else:
+            with open(str(f), 'rb') as fp:
+                data = fp.read(1024)
+                if is_binary_string(data):
+                    logging.debug(
+                        "[File contains {} bytes of binary data.]".format(
+                            f.stat().st_size))
+                else:
+                    fp.seek(0)
+                    for line in fp:
+                        line_str = line.decode(locale.getpreferredencoding(),
+                                               errors='backslashreplace')
+                        logging.debug(line_str.rstrip())
+        logging.debug(
+            "^^^^^^^^^^^^^^^^^^^^ {} ^^^^^^^^^^^^^^^^^^^^\n".format(f))
+
+
+def load_sw_over_spi(tmp_path, spiflash_path, sw_test_bin, spiflash_args=[]):
+    """ Use the spiflash utility to load software onto a device. """
+
+    log.info("Flashing device software from {} over SPI".format(
+        str(sw_test_bin)))
+
+    cmd_flash = [spiflash_path, '--input', sw_test_bin] + spiflash_args
+    p_flash = Process(cmd_flash, logdir=tmp_path, cwd=tmp_path)
+    p_flash.run()
+    p_flash.proc.wait(timeout=600)
+    assert p_flash.proc.returncode == 0
+
+    log.info("Device software flashed.")
+
+
+def find_in_files(file_objects,
+                  pattern,
+                  timeout,
+                  filter_func=None,
+                  from_start=False,
+                  wait_func=None):
+    """Find a pattern in a list of file objects (file descriptors)
+
+    See the documentation of match_line() for a description of |pattern|.
+
+    Args:
+        pattern: Pattern to search for in STDOUT or STDERR.
+        timeout: Timeout in seconds for |pattern| check.
+        filter_func: Function to filter the output with before applying
+            |pattern|. If none, the |default_filter_func| is used.
+        from_start: Search in all of STDOUT and STDERR since the beginning
+            of the execution.
+    Returns:
+        None if |pattern| was not found, the return value of match_line()
+        otherwise.
+    """
+
+    if timeout != None:
+        t_end = time.time() + timeout
+
+    if wait_func is None:
+        # By default, sleep for 200 ms when waiting for new input.
+        wait_func = lambda: time.sleep(.2)
+
+    if from_start:
+        for file_object in file_objects:
+            file_object.seek(0)
+
+    while True:
+        for file_object in file_objects:
+            i = 0
+            for line in file_object:
+                i += 1
+
+                m = match_line(line.rstrip(), pattern, filter_func)
+                if m is not None:
+                    return m
+
+        # Check if we exceed the timeout. Do so only every 100 lines to
+        # reduce the performance impact.
+        if timeout != None and i % 100 == 99 and time.time() >= t_end:
+            break
```

Comment:
This is clever! (And I didn't know you could iterate over a file object again after getting to EOF). Would it make sense to also check whether the process has finished? If so, there's no need to carry on waiting, right?